pyface/ui/qt4/font.py

```diff
@@ -0,0 +1,225 @@
+    QFont.System: 'default',
+    QFont.Decorative: 'fantasy',
+    QFont.Serif: 'serif',
+    QFont.Cursive: 'cursive',
+    QFont.SansSerif: 'sans-serif',
+    QFont.Monospace: 'monospace',
+    QFont.TypeWriter: 'typewriter',
+}
+generic_family_to_qt_family = {
+    'default': QFont.System,
+    'fantasy': QFont.Decorative,
+    'decorative': QFont.Decorative,
+    'serif': QFont.Serif,
+    'roman': QFont.Serif,
+    'cursive': QFont.Cursive,
+    'script': QFont.Cursive,
+    'sans-serif': QFont.SansSerif,
+    'swiss': QFont.SansSerif,
+    'monospace': QFont.Monospace,
+    'modern': QFont.Monospace,
+    'typewriter': QFont.TypeWriter,
+}
+
+weight_to_qt_weight = {
+    100: 0,
+    200: QFont.ExtraLight,
+    300: QFont.Light,
+    400: QFont.Normal,
+    500: QFont.Medium,
+    600: QFont.DemiBold,
+    700: QFont.Bold,
+    800: QFont.ExtraBold,
+    900: QFont.Black,
+    1000: 99,
+}
+qt_weight_to_weight = {
+    0: 'thin',
+    QFont.ExtraLight: 'extra-light',
+    QFont.Light: 'light',
+    QFont.Normal: 'normal',
+    QFont.Medium: 'medium',
+    QFont.DemiBold: 'demibold',
+    QFont.Bold: 'bold',
+    QFont.ExtraBold: 'extra-bold',
+    QFont.Black: 'black',
+    99: 'extra-heavy',
+}
+
+stretch_to_qt_stretch = {
+    100: QFont.UltraCondensed,
+    200: QFont.ExtraCondensed,
+    300: QFont.Condensed,
+    400: QFont.SemiCondensed,
+    500: QFont.Normal,
+    600: QFont.SemiExpanded,
+    700: QFont.Expanded,
+    800: QFont.ExtraExpanded,
+    900: QFont.UltraExpanded,
+}
+qt_stretch_to_stretch = {
+    QFont.UltraCondensed: 'ultra-condensed',
+    QFont.ExtraCondensed: 'extra-condensed',
+    QFont.Condensed: 'condensed',
+    QFont.SemiCondensed: 'semi-condensed',
+    QFont.Normal: 'normal',
+    QFont.SemiExpanded: 'semi-expanded',
+    QFont.Expanded: 'expanded',
+    QFont.ExtraExpanded: 'extra-expanded',
+    QFont.UltraExpanded: 'ultra-expanded',
+}
+
+style_to_qt_style = {
+    'normal': QFont.StyleNormal,
+    'oblique': QFont.StyleOblique,
+    'italic': QFont.StyleItalic,
+}
+qt_style_to_style = {value: key for key, value in style_to_qt_style.items()}
+
+
+def font_to_toolkit_font(font):
+    """ Convert a Pyface font to a Qfont.
+
+    Parameters
+    ----------
+    font : pyface.font.Font
+        The Pyface font to convert.
+
+    Returns
+    -------
+    qt_font : QFont
+        The best matching Qt font.
+    """
+    qt_font = QFont()
+    families = []
+    default_family = None
+
+    for family in font.family:
+        if family not in generic_family_to_qt_family:
+            families.append(family)
+        elif default_family is None:
+            default_family = family
+
+    if families and hasattr(qt_font, 'setFamilies'):
+        # Qt 5.13 and later
+        qt_font.setFamilies(families)
+    elif families:
+        qt_font.setFamily(families[0])
+        # Note: possibily could use substitutions here,
+        # but not sure if global (which would be bad, so we don't)
```

Comment:
This class (along with `QFont` and `wx.Font`) is a font _specification_, not a specific font instance.  What it provides is a way for a developer to state "I'm OK with something that looks like this" and then lets the toolkit and operating system sort out what the best fit is.

When it comes to Enable and Kiva font _rendering_ we will end up caring a lot about this, because for some backends we will have to find the "best fit" font so that we can draw it.  But that's not the point of this `Font` object.

The comment here is that Qt < 5.13 provides a way of specifying "it's OK to use this family rather than that family", but that I think it means messing with global state (ie. what you're specifiying is "it's _always_ OK to use this family rather than that family").  That's more wrong than just taking the first family option.

We definitely _don't_ want to create a `QApplication` as a side-effect.

The `px` problem is a separate issue and should be fixed.

Reply:
Agreed that we don't want to have the side-effect of creating `QApplication`.  In fact, that wasn't what I was suggesting to do. I think `QApplication.instance()` will return `None` if the `QApplication` has not been created, and in that case, definitely don't create a QApplication. But if the returned value is not None, then we can loop over the family instead of picking the first one which may not work (whereas the second/third/... might). That is a nice-to-have I think. At the very least, I think this branch could log a warning saying we are picking off the first one and ignoring the rest (if there are any).

Reply:
Just to be more specific about what I was thinking about. This does not have to happen here, but an idea for consideration:
```
def find_family(families): 
    if len(families) == 1: 
        return families[0] 
    app = QtGui.QApplication.instance() 
    if app is None: 
        logger.warning("using the first one only") 
        return families[0] 
    for family in families: 
        if QtGui.QFontInfo(QtGui.QFont(family)).family() == family: 
            return family 
    logger.warning("using the first one only after trying all.") 
    return families[0]
```
e.g.
```
# QApplication has not been created
>>> find_family(["Not Found", "Arial"])
using the first one only
'Not Found'
>>> QtGui.QApplication([])
>>> find_family(["Arial"])
"Arial" 
```
For consideration only as I have to agree that this is a bit magical, just to support something missing in Qt < 5.13.